crates/ruff_python_ast/src/source_code/line_index.rs

```diff
@@ -236,49 +236,71 @@ impl IndexKind {
 /// Type-safe wrapper for a value whose logical range starts at `1`, for
 /// instance the line or column numbers in a file
 ///
-/// Internally this is represented as a [`NonZeroUsize`], this enables some
+/// Internally this is represented as a [`NonZeroU32`], this enables some
 /// memory optimizations
+///
+/// Conversion between `usize` and `OneIndexed` always specify
+/// zero-indexed or one-indexed. The internal representation is always
+/// one-indexed.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub struct OneIndexed(NonZeroUsize);
+pub struct OneIndexed(NonZeroU32);
+
+#[allow(clippy::cast_possible_truncation)] // manually checked
+const fn index_to_u32(value: usize) -> u32 {
+    assert!(
+        value <= u32::MAX as usize,
+        "Value is too large to be represented as a OneIndexed"
+    );
+    value as u32
+}
 
 impl OneIndexed {
     // SAFETY: These constants are being initialized with non-zero values
     /// The smallest value that can be represented by this integer type.
     pub const MIN: Self = unwrap(Self::new(1));
     /// The largest value that can be represented by this integer type
-    pub const MAX: Self = unwrap(Self::new(usize::MAX));
+    pub const MAX: Self = unwrap(Self::new(u32::MAX));
 
-    pub const ONE: NonZeroUsize = unwrap(NonZeroUsize::new(1));
+    const ONE: NonZeroU32 = unwrap(NonZeroU32::new(1));
```

Comment:
This is unrelated to the described change. I suggest to remove `pub` here or rename `ONE` to something other like `NONZERO_ONE`. The name `ONE` makes it looks like `Self` type, but it isn't.